### PR TITLE
os module nodejs support

### PR DESCRIPTION
**Pull request by bung87**

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -878,24 +883,27 @@ include "includes/oserr"
...
+    ## * `setCurrentDir proc <#setCurrentDir,string>`_
+    runnableExamples:
+      assert getHomeDir() == expandTilde("~")
 
-  when defined(windows): return string(getEnv("USERPROFILE")) & "\\"
-  else: return string(getEnv("HOME")) & "/"
+    when defined(windows): return string(getEnv("USERPROFILE")) & "\\" 
+    else: return string(getEnv("HOME")) & "/"
+else:
+  proc getHomeDir*(): string {.inline.} = jsos.homedir()
```
Comment (alaviss) [inline lib/pure/os.nim:906]: You can just implement this and everything else as an another `when` branch instead of doing it like this.

Comment (timotheecour) [inline lib/pure/os.nim:906]: bug: this needs to be consistent with other procs and ensure `/` is added

**Comment by timotheecour** [comment]
* refs: https://github.com/nim-lang/Nim/pull/14512 (ie explains why doing the change here instead of pathutils)
* @bung87 I tried your PR, it doesn't work yet;
among other things you're hitting this: https://github.com/nim-lang/Nim/issues/13469

File: lib/pure/os.nim (hunk 1 above)
Comment (bung87) [reply lib/pure/os.nim:906]: ok

**Comment by bung87** [comment]
> * refs: #14512 (ie explains why doing the change here instead of pathutils)
> * @bung87 I tried your PR, it doesn't work yet;
>   among other things you're hitting this: #13469

Thanks , things to be more clear to me , I'll continue this later

**Comment by Araq** [comment]
We dont' support NodeJS and patches like these make ``os.nim`` harder to maintain. What we can add an ``os_nodejs.nim`` module to fusion though.

**Comment by bung87** [comment]
> We dont' support NodeJS and patches like these make `os.nim` harder to maintain. What we can add an `os_nodejs.nim` module to fusion though.

I agree, make the code has too much branch, feel free close this, I'll move these to other place.

**Comment by timotheecour** [comment]
> We dont' support NodeJS and patches like these make os.nim harder to maintain. What we can add an os_nodejs.nim module to fusion though.

that's a self full-filling prophecy; note that testament itself relies on `-d:nodejs` to run js tests, and it's generally easier to test on nodejs than on browser; but more importantly it's a platform that is useful to support, eg to ease interop with hosts of existing applications written for node. 

I think supporting nodejs is very natural, and can be done in a way that doesn't scar os.nim (that's my goal with https://github.com/nim-lang/Nim/pull/14543 which should help this PR)
Using `os_nodejs.nim` as in internal module imported by os.nim is fine, but exposing `os_nodejs.nim` for users IMO is not a good idea.

**Comment by dom96** [comment]
No, there is very little reason for the core team to support node js. And no matter how much you argue that it won't affect anything because it's in a different file: it's still a PR and needs reviews and wastes Nim core dev time. Stop wasting that time on things like this, please.

**Pull request closed without merging**